frame-reader.go

```diff
@@ -0,0 +1,150 @@
+// Read a generic "framed" packet consisting of a header and a
+// This is used for both TLS Records and TLS Handshake Messages
+package mint
+
+import (
+	"fmt"
+	"io"
+)
+
+type frameDetails interface {
+	headerLen() int
+	defaultReadLen() int
+	frameLen(hdr []byte) (int, error)
+}
+
+const (
+	kFrameReaderHdr  = 0
+	kFrameReaderBody = 1
+)
+
+var frameReaderWouldBlock = fmt.Errorf("Would have blocked")
+
+type frameNextAction func(f *frameReader) error
+
+type frameReader struct {
+	conn        io.Reader
```

Comment:
It seems like right now there are two ways for data to come into the `frameReader`, (1) from `conn`and (2) from `addChunk`.  It seems like you could potentially simplify this by getting rid of the `conn` option.  You would just be requiring the caller to explicitly pass chunks in from its reader.  But then you could probably simplify some of the logic here and fold it into addChunk (i.e., just have `addChunk` find all the complete frames in the extended byffer and either return them or buffer them).   Removing the `conn` option would also remove all the concerns around blocking.

```
for !fr.HasFrames() {
    in.Read(buf) // <-- wrap blocking logic around this
    fr.AddChunk(buf)
}
```